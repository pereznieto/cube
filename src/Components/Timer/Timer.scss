@import '../../styles/main';

.timer {
  margin: 30px;
  color: $primary;
  font-size: 180px;
  font-family: $timer-font;

  .tenths,
  .minutes,
  .hundredths {
    font-size: 140px;
  }

  &.running {
    color: rgba($primary, .7);

    .seconds {
      color: $primary;
    }

    .hundredths {
      display: none;
    }
  }
}
