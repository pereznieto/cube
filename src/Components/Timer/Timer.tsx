import React, { Component } from 'react';
import cx from 'classnames';
import { formatTime } from '../../utilities/formatters';
import styles from './Timer.scss';

interface Props {
  isRunning: boolean;
  isLoading: boolean;
  isReady: boolean;
  callback?: (count: number) => void;
}

interface State {
  count: number;
}

class Timer extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      count: 0,
    };
  }

  private interval: NodeJS.Timer;

  componentDidUpdate(prevProps: Props, prevState: State) {
    const { isRunning, callback } = this.props;
    const { count } = this.state;

    if (prevProps.isRunning !== isRunning) {
      if (isRunning) {
        if (count > 0) {
          this.setState({ count: 0 });
        }
        this.interval = setInterval(() => this._tick(), 10);
      } else {
        clearInterval(this.interval);
        callback && callback(count);
      }
    }
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  private _tick = () => {
    this.setState({ count: this.state.count + 1 });
  }

  render() {
    const { isRunning, isLoading, isReady } = this.props;
    const { count } = this.state;
    const [minutes, seconds, tenths, hundredths] = (isRunning || count > 0) ? formatTime(count) : [0, '00', 0, 0];

    return (
      <div
        className={cx(styles['timer'], {
          [`${styles['running']}`]: isRunning,
          [`${styles['loading']}`]: isLoading,
          [`${styles['ready']}`]: isReady,
        })}
      >
        <p className={styles['time']}>
          {minutes ? <span className={styles['minutes']}>{minutes}:</span> : null}
          <span className={styles['seconds']}>{seconds}</span>
          <span className={styles['dot']}>.</span>
          <span className={styles['tenths']}>{tenths}</span>
          <span className={styles['hundredths']}>{hundredths}</span>
        </p>
      </div>
    );
  }
}

export default Timer;
