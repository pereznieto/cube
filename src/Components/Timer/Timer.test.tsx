import { shallow } from 'enzyme';
import * as React from 'react';
import Timer from './Timer';

const props = {
  isRunning: true,
};

describe('Timer', () => {
  it('should render without crashing', () => {
    shallow(<Timer {...props} />);
  });

  it('should render Timer as expected', () => {
    const component = shallow(<Timer {...props} />);
    expect(component).toMatchSnapshot();
  });
});
