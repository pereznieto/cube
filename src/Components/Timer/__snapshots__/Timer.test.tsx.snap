// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should render Timer as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Timer
    isRunning={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p
        className={undefined}
      >
        <span
          className={undefined}
        >
          00
        </span>
        <span
          className={undefined}
        >
          .
        </span>
        <span
          className={undefined}
        >
          0
        </span>
        <span
          className={undefined}
        >
          0
        </span>
      </p>,
      "className": "undefined",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <span
            className={undefined}
          >
            00
          </span>,
          <span
            className={undefined}
          >
            .
          </span>,
          <span
            className={undefined}
          >
            0
          </span>,
          <span
            className={undefined}
          >
            0
          </span>,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "00",
            "className": undefined,
          },
          "ref": null,
          "rendered": "00",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": ".",
            "className": undefined,
          },
          "ref": null,
          "rendered": ".",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": undefined,
          },
          "ref": null,
          "rendered": 0,
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "0",
            "className": undefined,
          },
          "ref": null,
          "rendered": "0",
          "type": "span",
        },
      ],
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p
          className={undefined}
        >
          <span
            className={undefined}
          >
            00
          </span>
          <span
            className={undefined}
          >
            .
          </span>
          <span
            className={undefined}
          >
            0
          </span>
          <span
            className={undefined}
          >
            0
          </span>
        </p>,
        "className": "undefined",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            null,
            <span
              className={undefined}
            >
              00
            </span>,
            <span
              className={undefined}
            >
              .
            </span>,
            <span
              className={undefined}
            >
              0
            </span>,
            <span
              className={undefined}
            >
              0
            </span>,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "00",
              "className": undefined,
            },
            "ref": null,
            "rendered": "00",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": ".",
              "className": undefined,
            },
            "ref": null,
            "rendered": ".",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": undefined,
            },
            "ref": null,
            "rendered": 0,
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "0",
              "className": undefined,
            },
            "ref": null,
            "rendered": "0",
            "type": "span",
          },
        ],
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
